File: python-bindings/tests/game_setup_intake_conformance.py
"""Public Game Setup Intake using fully explicit isolated path inputs."""

import tempfile
from pathlib import Path


def observe_setup(fixture: dict) -> dict:
    """Compare direct and settings-backed intake without permitting ambient path discovery."""
    import classic_scangame

    if fixture["operation"] == "normalize":
        return {
            "versions": [
                classic_scangame.normalize_game_setup_version_selection(value)
                for value in fixture["versions"]
            ],
            "needs": [
                list(classic_scangame.game_setup_needs_path_detection(*paths))
                for paths in fixture["paths"]
            ],
        }
    with tempfile.TemporaryDirectory(prefix="classic-setup-conformance-") as temporary:
        root = Path(temporary)
        for name, content in fixture["files"].items():
            if (
                    "\\" in name
                    or Path(name).is_absolute()
                    or any(part in {"", ".", ".."} for part in name.split("/"))
            ):
                raise ValueError("setup fixture escaped root")
            path = root / name
            path.parent.mkdir(parents=True, exist_ok=True)
            path.write_bytes(content.replace("<ROOT>", root.as_posix()).encode())

        def tree():
            """Capture all owned files and directories before and after read-only intake."""
            return {
                path.relative_to(root).as_posix(): path.read_bytes()
                if path.is_file()
                else None
                for path in sorted(root.rglob("*"))
            }

        before = tree()
        intake = classic_scangame.GameSetupIntake(
            "Starfield",
            "Original",
            game_root=root / "Game",
            docs_root=root / "Docs",
            game_exe_path=root / "Game/Starfield.exe",
        )
        if (
                intake.game_id != "Starfield"
                or intake.game_version != "Original"
                or Path(intake.game_root) != root / "Game"
        ):
            raise ValueError("setup constructor lost supplied facts")

        def project(result):
            """Read stable summary fields and verify typed checks agree with rendered output."""
            report = result.combined()
            if (
                    report != result.rendered_report
                    or result.total_checks != len(result.checks)
                    or result.failed_checks
                    != sum(check.state == "failed" for check in result.checks)
            ):
                raise ValueError("setup summary lost typed diagnostic facts")
            for check in result.checks:
                if (
                        f"[{check.state}] {check.kind}: {check.message}" not in report
                        or any(detail not in report for detail in check.details)
                ):
                    raise ValueError("setup report lost diagnostic content")
            return {
                "status": result.status,
                "hasErrors": result.has_errors,
                "totalChecks": result.total_checks,
                "failedChecks": result.failed_checks,
                "actionCount": result.action_count,
                "pathUpdateCount": result.path_update_count,
                "pathUpdates": [
                    {
                        "kind": update.kind,
                        "path": Path(update.path).relative_to(root).as_posix(),
                    }
                    for update in result.path_updates
                ],
                "gameRoot": Path(result.game_root).relative_to(root).as_posix(),
                "docsRoot": Path(result.docs_root).relative_to(root).as_posix(),
                "gameExecutable": Path(result.game_executable)
                .relative_to(root)
                .as_posix(),
                "reportFlags": {
                    "gameNamed": "Game Setup Intake: Starfield" in report,
                    "metadataUnsupported": "[unsupported] registry_metadata:" in report,
                    "versionWarning": "[warning] executable_version:" in report,
                    "documentsPassed": "[passed] documents_folder:" in report,
                    "loaderFailed": "[failed] xse_loader:" in report,
                },
            }

        direct = project(classic_scangame.run_game_setup_intake(intake))
        result = project(
            classic_scangame.run_game_setup_intake_from_user_settings(root)
        )
        if result != direct:
            raise ValueError(
                "settings-backed intake differs from equivalent explicit facts"
            )
        after = tree()
        return {
            **result,
            "files": [name for name, content in after.items() if content is not None],
            "unchanged": before == after,
        }

File: node-bindings/classic-node/__test__/game_setup_intake_conformance.ts
import {mkdir, mkdtemp, readdir, readFile, rm, writeFile} from "node:fs/promises";
import {tmpdir} from "node:os";
import {dirname, isAbsolute, join, relative} from "node:path";
import * as classic from "../index.js";

/** Execute primitive decisions or compare explicit and settings-backed read-only setup intake. */
export async function observeSetup(fixture: any): Promise<Record<string, unknown>> {
    if (fixture.operation === "normalize") return {
        versions: fixture.versions.map((value: any) => classic.normalizeGameSetupVersionSelection(value)),
        needs: fixture.paths.map(([game, docs]: any[]) => {
            const result = classic.gameSetupNeedsPathDetection(game, docs);
            return [result.needsGamePath, result.needsDocsPath];
        })
    };
    const root = await mkdtemp(join(tmpdir(), "classic-setup-conformance-"));
    try {
        for (const [name, content] of Object.entries(fixture.files)) {
            if (name.includes("\\") || isAbsolute(name) || name.split("/").some(part => ["", ".", ".."].includes(part))) throw new Error("setup fixture escaped root");
            await mkdir(dirname(join(root, name)), {recursive: true});
            await writeFile(join(root, name), String(content).replaceAll("<ROOT>", root.replaceAll("\\", "/")));
        }

        /** Inventory all files/directories to catch unexpected write effects. */
        async function tree(): Promise<Record<string, string | null>> {
            const result: Record<string, string | null> = {};

            /** Capture nested owned entries without following links. */
            async function visit(dir: string): Promise<void> {
                for (const entry of await readdir(dir, {withFileTypes: true})) {
                    const path = join(dir, entry.name), key = relative(root, path).replaceAll("\\", "/");
                    if (entry.isSymbolicLink()) throw new Error("unexpected setup link");
                    if (entry.isDirectory()) {
                        result[key] = null;
                        await visit(path);
                    } else result[key] = (await readFile(path)).toString("hex");
                }
            }

            await visit(root);
            return Object.fromEntries(Object.entries(result).sort(([a], [b]) => a.localeCompare(b)));
        }

        const before = await tree();

        /** Preserve result summaries while authenticating rendered typed diagnostic fields. */
        function project(result: any): any {
            const report = result.renderedReport;
            if (result.totalChecks !== result.checks.length || result.failedChecks !== result.checks.filter((check: any) => check.state === "failed").length) throw new Error("setup summary lost typed checks");
            for (const check of result.checks) {
                if (!report.includes(`[${check.state}] ${check.kind}: ${check.message}`) || check.details.some((detail: string) => !report.includes(detail))) throw new Error("setup report lost diagnostic content");
            }
            return {
                status: result.status,
                hasErrors: result.hasErrors,
                totalChecks: result.totalChecks,
                failedChecks: result.failedChecks,
                actionCount: result.actionCount,
                pathUpdateCount: result.pathUpdateCount,
                pathUpdates: result.pathUpdates.map((update: any) => ({
                    kind: update.kind,
                    path: relative(root, update.path).replaceAll("\\", "/")
                })),
                gameRoot: relative(root, result.gameRoot).replaceAll("\\", "/"),
                docsRoot: relative(root, result.docsRoot).replaceAll("\\", "/"),
                gameExecutable: relative(root, result.gameExecutable).replaceAll("\\", "/"),
                reportFlags: {
                    gameNamed: report.includes("Game Setup Intake: Starfield"),
                    metadataUnsupported: report.includes("[unsupported] registry_metadata:"),
                    versionWarning: report.includes("[warning] executable_version:"),
                    documentsPassed: report.includes("[passed] documents_folder:"),
                    loaderFailed: report.includes("[failed] xse_loader:")
                }
            };
        }

        const direct = project(classic.runGameSetupIntake({
            gameId: "Starfield",
            gameVersion: "Original",
            gameRoot: join(root, "Game"),
            docsRoot: join(root, "Docs"),
            gameExePath: join(root, "Game/Starfield.exe")
        }));
        const result = project(classic.runGameSetupIntakeFromUserSettings(root));
        if (JSON.stringify(direct) !== JSON.stringify(result)) throw new Error("equivalent setup facts differ");
        const after = await tree();
        return {
            ...result,
            files: Object.entries(after).filter(([, value]) => value !== null).map(([path]) => path),
            unchanged: JSON.stringify(before) === JSON.stringify(after)
        };
    } finally {
        await rm(root, {recursive: true, force: true});
    }
}
